walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,111 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    // If duration is longer than a day, no option for this case.
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+      return Arrays.asList();
+    }
+    // Get general available time ranges.
+    Collection<String> attendees = request.getAttendees();
+    List<TimeRange> unavailableTime =
+        getNoOverlapTime(getAttendeesUnavailableTime(events, attendees));
+    List<TimeRange> availableTime = getAvailableTime(unavailableTime, request.getDuration());
+    return availableTime;
+  }
+
+  /** Get a collection of all attendees' unavailable time ranges (due to other events). */
+  private List<TimeRange> getAttendeesUnavailableTime(
+      Collection<Event> events, Collection<String> meetingAttendees) {
+    List<TimeRange> unavailableTime = new ArrayList<>();
+    for (Event event : events) {
+      Collection<String> eventAttendees = event.getAttendees();
+      // If any attendee in the meeting joins that event as well, the time range of the event is
+      // added to the unavailableTime collection.
+      if (!Collections.disjoint(meetingAttendees, eventAttendees)) {
+        unavailableTime.add(event.getWhen());
+      }
+    }
+    return unavailableTime;
+  }
+
+  /** Return a collection of time ranges in the time order without any overlap among each other. */
+  private List<TimeRange> getNoOverlapTime(List<TimeRange> unavailableTime) {
```

Comment:
Does this function assume the input range is sorted?

Reply:
No. The input of this function is the output of the helper function "getAttendeesUnavailableTime", so the input range is not sorted, and they may also have overlaps among the input time slots.
What "getNoOverlapTime" does is to merge the input time slots into sorted time without overlap.
Inside the function, we guarantee that time slots in "noOverlapTime" list are always sorted and merged without overlap. When a new time comes, it will be put into the correct sorted position and guaranteed to be merged correctly as well.